### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by amsowie**
PLAT 901, 902, and
removed the following dead code:
email_job.py
emailjob.py
entities reference to EmailSentTag in base_email.py
EmailJob reference in README
objects = EmailManager, line 31 in models.py
manager.py
EmailSentTag in models.py

### CODE REVIEW

File: emailengine/base_email.py [hunk 1]
```diff
@@ -157,14 +154,7 @@ def send(
                     attachments=attachments,
                 )
                 email_sent = self.record_send(send_to_user, send_to_email, email_job_record)
-
                 # Tag associated entities
```
Comment (RevolutionTech) [inline emailengine/base_email.py]: nit: Remove stale comment

File: emailengine/models.py [hunk 2]
```diff
@@ -57,15 +56,3 @@ class EmailJobRecord(models.Model):
     total_count = models.PositiveIntegerField(null=True)
     sent_count = models.PositiveIntegerField(null=False, default=0)
     skip_count = models.PositiveIntegerField(null=False, default=0)
-
-
-class EmailSentTag(models.Model):
```
Comment (RevolutionTech) [inline emailengine/models.py]: Because this model was deleted, we'll need to create a new migration that removes the model.

File: emailengine/tests/test_attachment.html [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE html>
```
Comment (RevolutionTech) [inline emailengine/tests/test_attachment.html]: nit: Place all test data into its own directory (my suggestion: `emailengine/tests/testdata`)

File: emailengine/tests/urls.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
```
Comment (RevolutionTech) [inline emailengine/tests/urls.py]: Is this file necessary?

File: emailengine/utils.py [hunk 5]
```diff
@@ -1,3 +1,6 @@
+# -*- coding: utf-8 -*-
+# from __future__ import unicode_literals
```
Comment (RevolutionTech) [inline emailengine/utils.py]: Should this be commented out?

File: setup.py [hunk 6]
```diff
@@ -1,10 +1,75 @@
...
+        from django.core.management import call_command
+
+        settings.configure(
+            DATABASES={
+                'default': {
+                    'NAME': ':memory:',
+                    'ENGINE': 'django.db.backends.sqlite3',
+                },
+            },
+            INSTALLED_APPS=(
```
Comment (RevolutionTech) [inline setup.py]: Are all of these necessary? I think there's quite a few in here that can be removed.

File: setup.py [hunk 7]
```diff
@@ -1,10 +1,75 @@
...
+                'django.contrib.admin',
+                'django.contrib.auth',
+                'django.contrib.contenttypes',
+                'django.contrib.messages',
+                'django.contrib.sessions',
+                'django_tables2',
+                'emailengine',
+                'emailengine.tests',
+            ),
+            TEMPLATES=[
```
Comment (RevolutionTech) [inline setup.py]: Are `TEMPLATES`, `MIDDLEWARE`, `MIDDLEWARE_CLASSES`, and `ROOT_URLCONF` all necessary? I don't know if EmailEngine should ever need these.

File: tox.ini [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+envlist =
+   py27
+   dj{18,19,110,111}
+
+[testenv]
+passenv = CI CIRCLECI CIRCLE_*
+commands =
+    pip install -q coverage codecov
+    pip install mock
+    pip install django_tables2
```
Comment (RevolutionTech) [inline tox.ini]: Does `django_tables2` need to be installed?

File: emailengine/tests/test_base_email.py [hunk 9]
```diff
@@ -0,0 +1,218 @@
+# # -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from datetime import datetime, timedelta
+from mock import mock
+import os
+
+from django.core import mail
+from django.test import TestCase
+from email.mime.text import MIMEText
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: nit: `email` imports should be with the stdlib imports above the `django` imports

File: emailengine/tests/test_base_email.py [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+from emailengine.models import EmailSent
+
+
+class BaseEmailTestCase(TestCase):
+    """
+    Verify BaseEmail is saving, sending, tracking etc
+    """
+
+    def setUp(self):
+        self.email_instance = BaseEmail()
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: Maybe it would be cleaner to create your own email class that uses `BaseEmail` as a parent class.
```python
class FooEmail(BaseEmail):

    base_template_dir = 'emailengine/'
    body_template = 'base.html'
    html_template = 'base.html'


class BaseEmailTestCase(TestCase):

    def setUp(self):
        self.email_instance = FooEmail()
```

File: emailengine/tests/test_base_email.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+            context={'subject': 'emailengine/base'},
+            send_to_user=None,
+            send_to_email='test_email@fakemail.com',
+            email_from='fake_sender@fake.com',
+            email_from_name=None,
+            cc_emails=None,
+            email_job_record=None,
+            force_send=None,
+            dryrun=False,
+            entities=[self.email_record],
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: Should `entities` be removed now?

File: emailengine/base_email.py [hunk 12]
```diff
@@ -157,14 +154,7 @@ def send(
...
-                if entities is not None:
```
Comment (RevolutionTech) [inline emailengine/base_email.py]: Since `entities` is no longer used, we should remove it from the method signature too.

File: emailengine/tests/test_base_email.py [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+        self.email_instance = BaseEmail()
+        self.email_instance.base_template_dir = 'emailengine/'
+        self.email_instance.body_template = 'base.html'
+        self.email_instance.html_template = 'base.html'
+
+    def tearDown(self):
+        self.email_instance.outbox = []
+
+    def test_send_to_email(self):
+        self.email_record = EmailSent(
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: What does this line and the one below do?

File: emailengine/tests/test_base_email.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+        )
+        self.assertIn('test_email@fakemail.com', mail.outbox[0].to[0])
+
+    def test_record_send_with_email_not_user(self):
+        """
+        Verify record_send works without passing send_to_user
+        """
+        self.email_instance.email_type = 'email_type'
+        email_sent_object = self.email_instance.record_send(
+            None, 'fake@fakemail.com'
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: nit: Place each method argument on its own line
```python
email_sent_object = self.email_instance.record_send(
    None,
    'fake@fakemail.com'
)
```

File: emailengine/tests/test_base_email.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+        self.assertEquals(email_sent_object.email, 'fake@fakemail.com')
+
+    def test_record_send_no_email_type(self):
+        """
+        Verify record_send returns None if there is no email_type
+        """
+        value = self.email_instance.record_send(None, 'fake@fakemail.com')
+
+        self.assertEqual(value, None)
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: nit: Use `assertIsNone()` instead

File: emailengine/tests/test_base_email.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        )
+        self.assertIn('1234', mail.outbox[0].body)
+        self.assertEqual(len(mail.outbox), 1)
+        self.assertIn('base', mail.outbox[0].subject)
+
+    def test_attachments(self):
+        """
+        Verify attachments passed in are actually attached to email
+        """
+        self.record_send = EmailSent.objects.create(
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: Why this line and L118?

File: emailengine/tests/test_base_email.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+            entities=None,
+            attachments=None,
+        )
+        self.assertEqual(was_sent, False)
+
+    def test_max_send_count(self):
+        """
+        Verify email address receives email if no send_to_user
+        """
+        self.email_instance.max_send_cnt = 1
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: What does this line do? Is `max_send_cnt` what's really being tested here?

File: emailengine/tests/test_base_email.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+            cc_emails=None,
+            email_job_record=None,
+            force_send=None,
+            dryrun=False,
+            entities=None,
+            attachments=None,
+        )
+        self.assertEqual(was_sent, True)
+
+    def test_unsubscribed_method(self):
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: Wouldn't the user be subscribed by default anyways? Maybe in this test we should unsubscribe the user first.

File: emailengine/tests/test_base_email.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+            'fake@fakemail.com'
+        )
+        self.assertEqual(was_unsubscribed, False)
+
+    @mock.patch('emailengine.base_email.BaseEmail.unsubscribe_url')
+    def test_append_unsubscribe(self, mock_unsubscribed):
+        """
+        Verify urls are appended in unsubscribe
+        """
+        self.email_instance.send(
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: Do we need to send an email to test this?

File: emailengine/tests/test_base_email_user_case.py [hunk 20]
```diff
@@ -0,0 +1,147 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from datetime import datetime, timedelta
+from django.contrib.auth.models import User
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email_user_case.py]: `datetime` is a stdlib, so should be grouped before `django` imports:
```python
from datetime import datetime, timedelta

from django.contrib.auth.models import User
...
```

File: emailengine/tests/test_base_email_user_case.py [hunk 21]
```diff
@@ -0,0 +1,147 @@
...
+    """
+
+    def setUp(self):
+        self.email_instance = BaseEmail()
+        self.email_instance.email_type = 'please respond'
+        self.email_instance.base_template_dir = 'emailengine/'
+        self.email_instance.body_template = 'base.html'
+        self.email_instance.html_template = 'base.html'
+
+        self.email_record = User.objects.create(
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email_user_case.py]: `email_record` is a bit of a misleading name. Maybe we should use `user` instead?

File: emailengine/tests/test_base_email_user_case.py [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+          email_from='fake_sender2@fake.com',
+          email_from_name='bobs',
+          cc_emails=None,
+          email_job_record=None,
+          force_send=None,
+          dryrun=False,
+          entities=None,
+          attachments=None,
+        )
+        self.assertEqual(was_sent2, False)
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email_user_case.py]: A good test case might be to check that after the interval passes, an email can be sent once again. To do this, you may want to change the `sent_datetime` in the `EmailSent` instance.

File: emailengine/tests/test_base_email_user_case.py [hunk 23]
```diff
@@ -0,0 +1,147 @@
...
+        """
+        Verify email_from_name is used for email, doesn't send
+        """
+        self.email_instance.max_frequency = timedelta(seconds=1)
+        was_sent = self.email_instance.send(
+            context={'subject': 'testing'},
+            send_to_user=self.email_record.user,
+            send_to_email=None,
+            email_from='fake_sender@fake.com',
+            email_from_name='bobs',
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email_user_case.py]: Why would `email_from_name` impact whether or not an email is sent?

File: emailengine/tests/test_models.py [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+
+from emailengine.models import EmailSent
+
+
+class EmailSentTestCase(TestCase):
+    """
+    Verify clean method raises error
+    """
+
+    def test_clean(self):
```
Comment (RevolutionTech) [inline emailengine/tests/test_models.py]: Maybe there should also be a test case where `EmailSent` does not raise a `ValidationError`.

File: emailengine/tests/test_utils.py [hunk 25]
```diff
@@ -0,0 +1,57 @@
+# # -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import json
+from mock import mock
```
Comment (RevolutionTech) [inline emailengine/tests/test_utils.py]: `mock` and `requests` should be imported in a section after `json` since `json` is part of the stdlib and the others are not. This could come after the `django` imports or within the same group.

File: emailengine/tests/test_utils.py [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+from requests.models import Response
+
+from django.test import TestCase
+
+from emailengine.utils import (
+    MailChimpError,
```
Comment (RevolutionTech) [inline emailengine/tests/test_utils.py]: nit: classes and functions imported do not need to each be on a single line:
```python
from foo import (
    a, b, bar, baz, c, d
)
```

File: emailengine/tests/test_utils.py [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+        """
+        mock_response = mock.Mock(spec=Response)
+        mock_response.content = json.dumps({'test': 'response'})
+        mock_response_post.return_value = mock_response
+        fake_email = 'testing@gmail.com'
+        fake_key = 'abc-def'
+        fake_listid = [123]
+        unsubscribe_user_from_mailchimp(fake_key, fake_listid, fake_email)
+
+        assert mock_response_post.called
```
Comment (RevolutionTech) [inline emailengine/tests/test_utils.py]: `mock.assert_called_with()` would be better, it's generally good to avoid `assert` statements where possible.

File: emailengine/tests/urls.py (hunk 4 above)
Comment (amsowie) [reply emailengine/tests/urls.py]: Not if I remove the root url conf in the setup :) oops

File: setup.py (hunk 7 above)
Comment (amsowie) [reply setup.py]: templates are required because of the Template() call in base_email. I removed the rest

File: tox.ini (hunk 8 above)
Comment (amsowie) [reply tox.ini]: yes, I run into a no module named django tables2 importerror if I don't have it there. Should it be somewhere else?

Comment (RevolutionTech) [reply tox.ini]: I think if you remove `django_tables2` from `INSTALLED_APPS`, then you should be able to remove this line.

File: emailengine/tests/test_base_email.py (hunk 13 above)
Comment (amsowie) [reply emailengine/tests/test_base_email.py]: Turns out EmailSent isn't really doing anything for this test, but the frequency line is setting up for an if/else conditional to be tested in send()

File: emailengine/tests/test_base_email.py (hunk 17 above)
Comment (amsowie) [reply emailengine/tests/test_base_email.py]: I'm not sure. If max_send_cnt is not set then the else conditional doesn't run for email=send_to_user. Are you saying I should name the test function something different? Or that I should test max_send_cnt differently?

File: emailengine/tests/test_base_email.py (hunk 13 above)
Comment (RevolutionTech) [reply emailengine/tests/test_base_email.py]: Is there a case handled by the second line that isn't handled by one of your other tests, such as `test_max_frequency()` for instance?

File: emailengine/tests/test_base_email.py (hunk 19 above)
Comment (amsowie) [reply emailengine/tests/test_base_email.py]: Maybe not- but another related question: why when I removed the send() are lines 158-159 are no longer covered in base_email.py?    
 `except Exception as e:
            self.logger.log(logging.ERROR, e, exc_info=True)`

File: emailengine/tests/test_base_email.py (hunk 17 above)
Comment (RevolutionTech) [reply emailengine/tests/test_base_email.py]: Oh okay, I see what you mean. It might be cleaner to cover that line along with `test_max_send_count_send_to_user()` (although maybe renaming it to `test_max_send_count()`). I understand that you're trying to separate the two cases into two different test methods, but what makes this test method confusing is that it's claiming to test `max_send_cnt`, although the primary functionality of that attribute is to _prevent_ sending multiple emails to a specific user/email address, although in this test we don't send multiple emails.

File: emailengine/tests/test_base_email.py (hunk 19 above)
Comment (RevolutionTech) [reply emailengine/tests/test_base_email.py]: That's interesting. It seems like when you covered those lines you had found a bug in EmailEngine. Those lines shouldn't ever be covered by tests in theory though, because they would only be executed if some unknown error occurred, as far as I can tell. These lines can be safely ignored from coverage with using `pragma: no cover` though since this is just defensive code:
```python
try:
    if dryrun is False:
        self.send_smtp(
            ...
except Exception as e:  # pragma: no cover
    self.logger.log(logging.ERROR, e, exc_info=True)

return True
```

Regardless, I'd be curious to understand how you were able to get that exception to be raised consistently.

File: emailengine/tests/test_models.py (hunk 24 above)
Comment (amsowie) [reply emailengine/tests/test_models.py]: I'm adding this test, but not sure what I should check- right now I have assertIsNone because the clean method doesn't return anything. Should I check something else?

Comment (RevolutionTech) [reply emailengine/tests/test_models.py]: Effectively what you want to assert is that the `ValidationError` is not raised in the other case. There isn't an assertion method for that, so you can simply call the `clean()` method in the test (though a comment may make it clearer what you're doing):
```python
with self.assertRaises(ValidationError):
    self.record_send3 = ...
    self.record_send3.clean()

self.record_send4 = ...
self.record_send4.clean()  # does not raise ValidationError
```

File: emailengine/tests/test_base_email.py (hunk 17 above)
Comment (amsowie) [reply emailengine/tests/test_base_email.py]: Except the way I understand it, if no send_to_user is specified (and in this case it isn't)  `if send_to_user:
                cnt_qs = cnt_qs.filter(user_id=send_to_user.user.id)
            else:
                cnt_qs = cnt_qs.filter(email=send_to_user)`
 actually returns None to the next line. I think it's misleading, but maybe I'm not understanding what the code is doing. In the case of email=send_to_user, I'm actually testing that max_send_cnt isn't exceeded because it can't be with the user set to None. I could rename it test_max_send_count_no_user(). I also don't think I can cover those lines in the same test because the if/else is whether or not send_to_user is T/F.

File: emailengine/tests/test_base_email.py (hunk 19 above)
Comment (amsowie) [reply emailengine/tests/test_base_email.py]: I'm not sure what made it think the lines were covered, but when that send was in there, this printed to the console mid tests: "No handlers could be found for logger "emailengine"". Now it doesn't. I removed and added pragma.

File: emailengine/tests/test_base_email.py (hunk 13 above)
Comment (RevolutionTech) [reply emailengine/tests/test_base_email.py]: I see. I think it might be clearer to have a more full test here, since in this test we're not really testing the functionality of `max_frequency` -- we're just trying to cover a line that filters emails sent, but the name suggests otherwise.

This test could mirror the `test_max_frequency()`, for example (just using `send_to_email` instead of `send_to_user`).

File: emailengine/tests/test_base_email.py (hunk 17 above)
Comment (RevolutionTech) [reply emailengine/tests/test_base_email.py]: I think I am confused, are you saying the line `cnt = cnt_qs.count()` will set `cnt` to `None`? Why would any outcome of the lines above cause that?

The two different test cases you're describing don't need to be in the same test, it was just a suggestion if it was difficult to come up with a unique name to test the two cases. Regardless of whether they are separated or not, I can't think of a reason why that would impact the test coverage, as long as we still test all of the same edge cases. In other words, that's true that you can only cover one of the two lines in the `if`/`else` when calling `send()`, but you can call `send()` multiple times in the same test method to cover every case.

File: emailengine/tests/test_base_email.py (hunk 19 above)
Comment (RevolutionTech) [reply emailengine/tests/test_base_email.py]: Oh okay, that exception could be prevented by adding a `LOGGING` setting to `settings.configure()`. Nevertheless it remains a mystery of why an exception got raised to log an error in the first place. Oh well. 😛

File: emailengine/tests/test_utils.py [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+
+    @mock.patch('emailengine.utils.requests.post')
+    def test_unsubscribe_user_from_mailchimp(self, mock_response_post):
+        """
+        Verify call is made to mailchimp api
+        """
+        mock_response = mock.Mock(spec=Response)
+        mock_response.content = json.dumps({'test': 'response'})
+        mock_response_post.return_value = mock_response
+        fake_email = 'testing@gmail.com'
```
Comment (RevolutionTech) [inline emailengine/tests/test_utils.py]: nit: Add a blank newline after the mock prep to separate the mock prep code from the rest of the test

File: emailengine/tests/test_utils.py [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+        fake_key = 'abc-def'
+        fake_listid = [123]
+        unsubscribe_user_from_mailchimp(fake_key, fake_listid, fake_email)
+
+        mock_response_post.call_args(
+            'https://def.api.mailchimp.com/2.0/lists/unsubscribe.json',
+            data=('{"send_notify": false, "send_goodbye": false,"'
```
Comment (RevolutionTech) [inline emailengine/tests/test_utils.py]: This might be slightly more readable if you create a Python dictionary and then apply `json.dumps()` to it.

File: emailengine/tests/test_utils.py [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+    @mock.patch('emailengine.utils.requests.post')
+    def test_unsubscribe_user_from_mailchimp(self, mock_response_post):
+        """
+        Verify call is made to mailchimp api
+        """
+        mock_response = mock.Mock(spec=Response)
+        mock_response.content = json.dumps({'test': 'response'})
+        mock_response_post.return_value = mock_response
+        fake_email = 'testing@gmail.com'
+        fake_key = 'abc-def'
```
Comment (RevolutionTech) [inline emailengine/tests/test_utils.py]: MailChimp keys end in a region code, like `-us1`, so using a fakekey that ends in this way, although a subtle detail, might make this test slightly clearer:
```python
fake_key = abc123-us1
```

File: emailengine/tests/test_utils.py [hunk 31]
```diff
@@ -0,0 +1,60 @@
...
+            'https://def.api.mailchimp.com/2.0/lists/unsubscribe.json',
+            data=('{"send_notify": false, "send_goodbye": false,"'
+            '"apikey": "abc-def", "id": [123],'
+            '"email": {"email": "testing@gmail.com"}}')
+        )
+
+    @mock.patch('emailengine.utils.requests.post')
+    def test_unsubscribe_user_mailchimp_error(self, mock_response_post):
+        """
+        Verify call is made to mailchimp api and error produced
```
Comment (RevolutionTech) [inline emailengine/tests/test_utils.py]: Maybe a description like "Verify errors from MailChimp are propagated" would be a better description, since you're not really testing that MailChimp produces errors with this test.

File: emailengine/tests/test_base_email.py [hunk 32]
```diff
@@ -0,0 +1,270 @@
...
+        """
+        class NoEmailType(TestEmail):
+            email_type = None
+
+        email_instance = NoEmailType()
+
+        value = email_instance.record_send(None, 'fake@fakemail.com')
+        num_emails_sent = EmailSent.objects.count()
+
+        self.assertEqual(EmailSent.objects.count(), num_emails_sent)
```
[the remark this answers is not in the document]
Comment (amsowie) [reply emailengine/tests/test_base_email.py]: I removed the num_emails_sent line and am checking count against 1 (from the setup email). Is that more meaningful?

File: emailengine/tests/test_base_email.py [hunk 33]
```diff
@@ -0,0 +1,270 @@
...
+        """
+        class NoEmailType(TestEmail):
+            email_type = None
+
+        email_instance = NoEmailType()
+
+        value = email_instance.record_send(None, 'fake@fakemail.com')
+
+        # check that email object count remains 1, from setUp
+        self.assertEqual(EmailSent.objects.count(), 1)
```
Comment (RevolutionTech) [inline emailengine/tests/test_base_email.py]: This is good, but it might be clearer to get the number of emails sent before `record_send()` is called and storing that in a variable, then you can check if `EmailSent.objects.count()` is still equal to that variable here, after `record_send()` is called.

The benefit of doing it this way is that if another email gets sent in the `setUp()`, it won't break this test.
